### DISCUSSION TITLE
Dynamic filtering on static sites

### DISCUSSION THREAD

**Discussion by andrewfairlie** [vercel/next.js#16997] [Help]
This post is looking to confirm some initial understandings of Next, and query a potential workaround if our understanding is correct.

---

We're thinking about how we can make static pages but with dynamic filters (think eCommerce filters).

Let's say we wanted these filters...

| Key    | Values           |
|--------|------------------|
| Size   | Large            |
| Sleeve | Long      |
| Colour | Red, Blue, Green |

Traditionally we might use query parameters for this, with something like

```
shop.com/products?size=large&sleeve=long&colour=red|blue|green
```

Am I right in thinking that this doesn't work with static pages using getStaticProps? And we'd need to use SSR to use that kind of feature?

If that's the case, has anybody tried having a dynamic route of something like `/pages/products/[query]`, using that URL segment to emulate query parameters? We'd need to designate certain characters to deliminate the parts of the query. For example, we could use `!` to indicate new key, and `:` to split the key and its values. The URL might become

```
/pages/product/size:large!sleeve:long!colour:red,blue,green
```

In JS we could deconstruct that in `getStaticProps` and do whatever API/CMS logic is needed to get the filtered content, and then pass that onto the page's component as a Next prop...

```
  export async function getStaticProps({ params }) {
      const queryParam = params.query;
      const sections = queryParam.split("!");
      const formattedQueryParams = {};

      for(let section of sections){
        const key = section.split(":")[0];
        const values = section.split(":")[1];
        formattedQueryParams[key] = values.split(',');
      }



    // Pass post data to the page via props
    return {
      props: {
          query: formattedQueryParams
      },
      revalidate: 1,
    }
  }
```

Is this a sensible way to approach it? Or is it too clever of its own good, am I missing an easier way around this?

I appreciate any help with this, especially as it's an abstract thought rather than an immediate real-world problem.

**Comment by luisdiaslima** [comment]
same doubt here!

**Reply by andrewfairlie** [reply]
Hi @luisdiaslima in the end we've used getServerSideProps with code like this (TypeScript, but can be adapted into JS)

```
export const getServerSideProps: GetServerSideProps = async (context) => {
  const { query, locale, res } = context
  res.setHeader('Cache-Control', `s-maxage=600, stale-while-revalidate`)
  [....]
```

This seems to be the best of both worlds that we've found because you can use query params but you can also cache it with stale-while-revalidate so that line makes it feel pretty close to a static page, but of course this requires a server (we just use Vercel).

I hope that helps?

**Reply by luisdiaslima** [reply]
@andrewfairlie nice, i think i understood how it would look, thanks for the help! Would you have any more in-depth content on this for me?

**Comment by michalpulpan** [comment]
> This post is looking to confirm some initial understandings of Next, and query a potential workaround if our understanding is correct.
> 
> We're thinking about how we can make static pages but with dynamic filters (think eCommerce filters).
> 
> Let's say we wanted these filters...
> 
> Key	Values
> Size	Large
> Sleeve	Long
> Colour	Red, Blue, Green
> Traditionally we might use query parameters for this, with something like
> 
> ```
> shop.com/products?size=large&sleeve=long&colour=red|blue|green
> ```
> 
> Am I right in thinking that this doesn't work with static pages using getStaticProps? And we'd need to use SSR to use that kind of feature?
> 
> If that's the case, has anybody tried having a dynamic route of something like `/pages/products/[query]`, using that URL segment to emulate query parameters? We'd need to designate certain characters to deliminate the parts of the query. For example, we could use `!` to indicate new key, and `:` to split the key and its values. The URL might become
> 
> ```
> /pages/product/size:large!sleeve:long!colour:red,blue,green
> ```
> 
> In JS we could deconstruct that in `getStaticProps` and do whatever API/CMS logic is needed to get the filtered content, and then pass that onto the page's component as a Next prop...
> 
> ```
>   export async function getStaticProps({ params }) {
>       const queryParam = params.query;
>       const sections = queryParam.split("!");
>       const formattedQueryParams = {};
> 
>       for(let section of sections){
>         const key = section.split(":")[0];
>         const values = section.split(":")[1];
>         formattedQueryParams[key] = values.split(',');
>       }
> 
> 
> 
>     // Pass post data to the page via props
>     return {
>       props: {
>           query: formattedQueryParams
>       },
>       revalidate: 1,
>     }
>   }
> ```
> 
> Is this a sensible way to approach it? Or is it too clever of its own good, am I missing an easier way around this?
> 
> I appreciate any help with this, especially as it's an abstract thought rather than an immediate real-world problem.

This actually looks like something most big ecommerce players now days implement. Is there any update about it?

**Comment by RenatoCesarF** [comment]
I'm having the same doubt.

I have a [blog page](https://github.com/RenatoCesarF/codingideas) written in Next as a static page. My blog has many Categories and I wish I could filter the posts by a certain category (or categories). But I want to keep it as a static website.

I implemented a filter List[ in another page](https://codingideas.vercel.app/team) of it. Where I display a List of Teammates but is only 3 cards that are filtered and the useState already "don't like it". Also implemented a way to see all the [posts from a certain Category](https://codingideas.vercel.app/blog/tag/blog), but you couldn't filter for more than one Category.

I found[ this video](https://www.youtube.com/watch?v=b7XLYkqJ5uY) as well but doesn't seem to be a static page.

I think generating all the possibilities is not worth it. Mainly if you have multiples filters, it's something exponential
